core/src/main/java/org/apache/calcite/rel/rules/PruneEmptyRules.java

```diff
@@ -551,17 +529,15 @@ public interface JoinRightEmptyRuleConfig extends PruneEmptyRule.Config {
             // "select * from emp left join dept" is not necessarily empty if
             // dept is empty
             // join can be removed and take the left branch, all columns come from dept are null
-            List<RexNode> projects = new ArrayList<>(
-                left.getRowType().getFieldCount() + empty.getRowType().getFieldCount());
-            List<String> columnNames = new ArrayList<>(
-                left.getRowType().getFieldCount() + empty.getRowType().getFieldCount());
-            // left
-            copyProjects(
-                rexBuilder, left.getRowType(), join.getRowType(), 0, projects, columnNames);
-            // right
-            addNullLiterals(rexBuilder, empty, projects, columnNames);
-
-            RelNode project = call.builder().push(left).project(projects, columnNames).build();
+            List<Pair<? extends RexNode, String>> projects =
+                Stream.concat(
+                    getProjectStream(left.getRowType(), join.getRowType(), 0, rexBuilder),
+                    getNullLiteralStream(empty, rexBuilder)).collect(Collectors.toList());
+
+            RelNode project = call.builder()
+                .push(left)
+                .project(Pair.left(projects), Pair.right(projects))
+                .build();
```

Comment:
I believe that the field names are the same before and after. If so, you don't need to create pairs of (RexNode, String), just RexNode. Which I think would simplify getProjectStream and getNullLiteralStream.

Reply:
Sorry, just read your comment, let me try it.

Reply:
The field names can be inherited from the left branch but it has to be set for the `NULL` literals otherwise names like `$f9` are generated.
explected
```
LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8], DEPTNO0=[$9], NAME=[$10])
  LogicalProject(EMPNO=[CAST($0):INTEGER], ENAME=[CAST($1):VARCHAR(20)], JOB=[CAST($2):VARCHAR(10)], MGR=[$3], HIREDATE=[CAST($4):TIMESTAMP(0)], SAL=[CAST($5):INTEGER], COMM=[CAST($6):INTEGER], DEPTNO=[CAST($7):INTEGER], SLACKER=[CAST($8):BOOLEAN], DEPTNO0=[null:INTEGER], NAME=[null:VARCHAR(10)])
    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```
actual
```
LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8], DEPTNO0=[$9], NAME=[$10])
  LogicalProject(EMPNO=[CAST($0):INTEGER], ENAME=[CAST($1):VARCHAR(20)], JOB=[CAST($2):VARCHAR(10)], MGR=[$3], HIREDATE=[CAST($4):TIMESTAMP(0)], SAL=[CAST($5):INTEGER], COMM=[CAST($6):INTEGER], DEPTNO=[CAST($7):INTEGER], SLACKER=[CAST($8):BOOLEAN], $f9=[null:INTEGER], $f10=[null:VARCHAR(10)])
    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```

Reply:
It works, thanks for the hint.

Reply:
I have a feeling that it could be even shorter if you use Guava's `List.concat()` rather than `Stream.concat()`, and use `RelBuilder.fields()`. You'll have to assign the `RelBuilder` to a variable.